geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/MonitoringFeatureReader.scala

```diff
@@ -0,0 +1,58 @@
+package org.locationtech.geomesa.utils.geotools
+
+import java.util.concurrent.atomic.AtomicBoolean
+
+import org.geotools.data.store.ContentFeatureSource
+import org.geotools.data.{DelegatingFeatureReader, FeatureReader, Query}
+import org.geotools.filter.text.ecql.ECQL
+import org.locationtech.geomesa.utils.stats.{MethodProfiling, TimingsImpl}
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+
+class MonitoringFeatureReader(storeType: String, query: Query, delegate: FeatureReader[SimpleFeatureType, SimpleFeature])
+  extends DelegatingFeatureReader[SimpleFeatureType, SimpleFeature] with MethodProfiling {
+
+  protected var counter = 0L
+
+  private val closed = new AtomicBoolean(false)
+
+  implicit val timings = new TimingsImpl
+
+  override def getDelegate: FeatureReader[SimpleFeatureType, SimpleFeature] = delegate
+  override def getFeatureType: SimpleFeatureType = delegate.getFeatureType
+
+  override def next(): SimpleFeature = {
+    counter += 1
+    profile(delegate.next(), "next")
+  }
+  override def hasNext: Boolean = profile(delegate.hasNext, "hasNext")
+
+  override def close() = if (!closed.getAndSet(true)) {
+    closeOnce()
+  }
+
+  protected def closeOnce(): Unit = {
+
+    val stat = GeneralUsageStat(storeType,
+      getFeatureType.getTypeName,
+      System.currentTimeMillis(),
+      ECQL.toCQL(query.getFilter),
+      timings.time("next") + timings.time("hasNext"),
+      counter
+    )
+    delegate.close()
+    Monitoring.log(stat)
+  }
+}
+
+trait MonitoringFeatureSourceSupport extends ContentFeatureSourceSupport {
+  self: ContentFeatureSource =>      // JNH: Do we need the self type?
```

Comment:
I would have thought that, if anywhere, the self type belongs in the ```ContentFeatureSourceSupport``` class rather than here.